### ISSUE TITLE
Thinkpad Clickpad middle-button "spams" middle-clicks when scrolling with TrackPoint

### ISSUE THREAD

**Issue by yattoz** [acidanthera/bugtracker#1566]
Hello, I came across a somewhat annoying bug with my Thinkpad Clickpad (on Thinkpad T470). 
I am using VoodooPS2Controller.kext and SSDT_Thinkpad_Clickpad.aml. All my input methods are PS/2.

On the Clickpad there are 3 buttons, on windows they're acting like left click, scroll aid, and right click. The middle button, "scroll aid", is not a middle click button but it has as scrolling function when it's hold down, scrolling then works by tilting the TrackPoint (that red dot on the keyboard).

For some reason, on macOS, the middle button is used as middle click. It's not a problem *per se*, but I get this odd behaviour:

What I do:
- I hold down the middle button of the Clickpad
- I tilt the TrackPoint down or up on a webpage

What I expect:
- I expect the page to scroll naturally without any clicks

What I get:
- The page does scroll but when doing so, the Clickpad sends middle-click events repeatedly (probably at every frame? I'm not sure). As a result, when scrolling a webpage, the mouse comes over links that are repeatedly middle-clicked, which open a shitton of new tabs.

I don't know really when there problem comes from, I read the Thinkpad_Clickpad DSL but it simply overrides some values from Synaptics, which I assume is the touchpad part. So I don't know how to deal with the behaviour of the click buttons.

Do you know where this "middle-click-spamming" behaviour comes from, and if it's possible to fix it? 
Thank you for all the hard word!

**Comment by vit9696** [comment]
I do not think we support Trackpoints, but maybe @usr-sse2 knows more.

**Comment by usr-sse2** [comment]
@vit9696 We do support trackpoints, although I don't have the hardware to test on.
Another reason to implement packet dumping and simulation.

**Comment by usr-sse2** [comment]
There is code in VoodooPS2SynapticsTouchpad that combines buttons from both sources (touchpad and trackpoint), if something is incorrect there, such behavior can happen.

**Comment by vit9696** [comment]
Ok, but then two questions:

- Can we guarantee the system is configured correctly and this issue refers to the issue in our code? Does trackpad support need any configuration?
- Can we guarantee that this exact trackpad does provide full trackpoint support over PS/2? I am pretty sure at least some trackpads provide full features only when working over SMBus.

**Comment by yattoz** [comment]
From what I see :

- The behaviour "use middle button as middle-click" has been designed as such in VoodooPS2SynapticTouchpad.cpp (line 904): 

https://github.com/acidanthera/VoodooPS2/blob/ab49b27d22b0de72677bdb44bd0e304d6495afde/VoodooPS2Trackpad/VoodooPS2SynapticsTouchPad.cpp#L904

In the same file, line 925, it mentions differenciating "middle click and middle scrolling" too, although I don't really get what these addresses mean.

The section that deals with middle-clicking and scrolling is around line 1024.

https://github.com/acidanthera/VoodooPS2/blob/ab49b27d22b0de72677bdb44bd0e304d6495afde/VoodooPS2Trackpad/VoodooPS2SynapticsTouchPad.cpp#L1024

I am still unsure how much time and effort I can give to check what's going on (especially since it's my very first time compiling anything on macOS, let alone dealing with hardware drivers), but if I have some spare time I could try to give it a shot and try to compile VoodooPS2 on my laptop. 

For my model, on Windows, the middle button acts as "scroll aid" ; but maybe for other models it acts as a middle click... or maybe the person who implemented that wasn't  That may explain why it has been implemented that way in VoodooPS2. 

If you know the code around these sections, and if you see somethign that could explain why middle-click events are repeatedly sent while scrolling, it would be much appreciated.

**Comment by 1Revenger1** [comment]
I had similar issues with VoodooRMI
Best way to handle it is to mask out the middle button and send it as a seperate button event when you release the middle button and haven't scrolled. Likewise, if the user scrolls, the middle button should never be sent once it's released. Fairly certain this is the behaviour that windows uses. 

I think it's a logic issue looking in VPS2 real quick, but I haven't actually tried it out.

**Comment by yattoz** [comment]
After compiling the code, taking a good look of how it works and what the logs reported, **I finally found out that it was a problem of configuration on my side.**

I was using another SSDT called `SSDT-HP-FixLidSleep` that I used at first to fix the problem of repeated key presses when I close the lid of my laptop (although it was not an HP at all, but I was advised to use that). Problem is, using this SSDT defines `_SB.PCI0.LPCB.KBD.RMCF`. Somehow the `SSDT-Thinkpad_Clickpad` is loaded after that, and fails to load since KBD.RMCF is already defined. As a result, when VoodooPS2Controller.cpp loads the ACPI values in `KBD.RMCF`, it didn't read anything, and my Trackpad wasn't recognized as a "Thinkpad Clickpad"... so in `VoodooPS2SynapticTouchpad.cpp` the variable `bool isthinkpad` was set to false, forbidding the correct code to be run.

After removing the `SSDT-HP-FixLidSleep` and writing the correct path for `_SB.PCI0.LPCB.KBD.RMCF` in  `SSDT-Thinkpad_Clickpad`, that SSDT loads correctly. Using `sudo dmseg | grep "ps2"` I can now see that logs reports `IS THINKPAD`. The scrolling problem disappears now, the middle button works as a "scroll aid" only and it never does any middle click (which is the default behaviour I was expecting, so it's good). The problem that I had when the keyboard was spamming keys right after sleeping doesn't appear either with this SSDT so it's all good, I guess.

It wasn't particularly easy to debug that but you made me ask myself the right questions, so we can say: case closed! Thank you very much.
